### PR TITLE
Alerting: Add limits to the Prometheus Rules API

### LINKED ISSUE #65420: Add limits to the Alert Rules endpoint `api/prometheus/grafana/api/v1/rules`
To limit the amount of data returned from the API, we nee to add limits the three main entities being returned from this endpoint:

- Rule Groups
- Rules within a Group
- Alerts

It's important to note that we want to preserve existing semantics and avoid a breaking change when possible - this means that we need to provide a way to ensure no pagination is the default

### PR DESCRIPTION
**Pull request by grobinson-grafana**
**What is this feature?**

This commit adds a number of limits to the Grafana flavor of the Prometheus Rules API:

1. `limit` limits the maximum number of Rule Groups returned
2. `limit_rules` limits the maximum number of rules per Rule Group
3. `limit_alerts` limits the maximum number of alerts per rule

It sorts Rule Groups and rules within Rule Groups such that data in the response is stable across requests.

It also returns summaries (totals) for all Rule Groups, individual Rule Groups and rules.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #65420, #65421, #65422, #65423

Epic https://github.com/grafana/grafana/issues/65071

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

### CODE REVIEW

File: pkg/services/ngalert/models/alert_rule.go [hunk 1]
```diff
@@ -173,6 +173,19 @@ type AlertRuleWithOptionals struct {
 	HasPause bool
 }
 
+// SortedAlertRules sorts alert rules by AlertRuleGroupKey and then their index.
+type SortedAlertRules []*AlertRule
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/models/alert_rule.go:177]: We already have this https://github.com/grafana/grafana/blob/f066e8cdcd178b783d0c62959bea593d3c0c30b0/pkg/services/ngalert/models/alert_rule.go#L496-L505
in the very same file. Although, I think the more explicit sorting function is more obvious, we can replace it with this.

Comment (grobinson-grafana) [reply pkg/services/ngalert/models/alert_rule.go:177]: Thank you! I didn't see this when I wrote that code :)

File: pkg/services/ngalert/api/api_prometheus.go [hunk 2]
```diff
@@ -115,6 +122,19 @@ func (srv PrometheusSrv) RouteGetRuleStatuses(c *contextmodel.ReqContext) respon
 		return ErrResp(http.StatusBadRequest, errors.New("panel_id must be set with dashboard_uid"), "")
 	}
 
+	limitGroups, err := getInt64FromRequest(c.Req, "limit")
```
Comment (gotjosh) [inline pkg/services/ngalert/api/api_prometheus.go:125]:
```suggestion
	limitGroups, err := c.QueryInt64("limit")
```

Not sure if you know, but Grafana already has a set of methods that do this modulo the trimming of spaces (e.g. `c.QueryInt64()`. Is the trimming of spaces required? In the past, the technique I've used is to limit back to the default limit if we're unable to do the conversion - so, something like:

```
func (ctx *ReqContext) QueryInt64WithDefault(field string, d int64) int64 {
	n, err := strconv.ParseInt(ctx.Query(field), 10, 64)
	if err != nil {
		return d
	}
	return n
}
```

I'm not marking this as a nit, as I do think we want "pagination" by default, even though it's technically a breaking change. Perhaps we could have a special case to _not_ have pagination e.g. `-1`

File: pkg/services/ngalert/models/alert_rule.go [hunk 3]
```diff
@@ -288,6 +301,16 @@ type AlertRuleGroupKey struct {
 	RuleGroup    string
 }
 
+func (k AlertRuleGroupKey) Less(v AlertRuleGroupKey) bool {
```
Comment (gotjosh) [inline pkg/services/ngalert/models/alert_rule.go:304]: Is this the sorting that the front-end expects? This is sorting by `IDs` as opposed to alphabetically, which is what the front-end currently expects.

File: pkg/services/ngalert/models/alert_rule.go (hunk 1 above)
Comment (gotjosh) [reply pkg/services/ngalert/models/alert_rule.go:177]: Not sure what's the decision to make here - and sorry if my comment is obvious, but they're technically not `SortedAlertRules` they're more of `SorteableAlertRules` as you'd still need to call sort on them to be sorted.

One more question though, why not include this directly in `ngmodels.AlertRule` is there something I'm missing?

**Review by gotjosh — DISMISSED** [review]
LGTM, but please see my comments.

**Review by JohnnyQQQQ — COMMENTED** [review]
Do we want to add the total number of groups/rules/instances to the api model?

File: pkg/services/ngalert/models/alert_rule.go (hunk 1 above)
Comment (grobinson-grafana) [reply pkg/services/ngalert/models/alert_rule.go:177]: I agree, will fix the name! I'm not sure I understand the last question though. How would we include a sorted slice of `[]AlertRule` in `ngmodels.AlertRule`?

File: pkg/services/ngalert/api/api_prometheus.go (hunk 2 above)
Comment (grobinson-grafana) [reply pkg/services/ngalert/api/api_prometheus.go:125]: `c.QueryInt64()` is problematic for what we need to do here 🙁 It cannot differentiate between a valid value of `0`, an invalid value, and a missing value.

File: pkg/services/ngalert/models/alert_rule.go (hunk 3 above)
Comment (grobinson-grafana) [reply pkg/services/ngalert/models/alert_rule.go:304]: Where do you see `ID` Josh? Here `AlertRuleGroupKey` sorts on Org, Namespace UID and then Rule Group?

Comment (grobinson-grafana) [reply pkg/services/ngalert/models/alert_rule.go:304]: Where do you see `ID` Josh? Here `AlertRuleGroupKey` sorts on Org ID, Namespace UID and then Rule Group?

**Comment by grobinson-grafana** [comment]
> Do we want to add the total number of groups/rules/instances to the api model?

Yes! 😄 I need to do it this morning 🚀

File: pkg/services/ngalert/models/alert_rule.go (hunk 3 above)
Comment (gotjosh) [reply pkg/services/ngalert/models/alert_rule.go:306]: I have resolved my own comment here.

**Review by gotjosh — APPROVED** [review]
LGTM

**Comment by grobinson-grafana** [comment]
> The backport to `v9.4.x` failed:
> 
> ```
> The process '/usr/bin/git' failed with exit code 1
> ```
> 
> To backport manually, run these commands in your terminal:
> 
> ```shell
> # Fetch latest updates from GitHub
> git fetch
> # Create a new branch
> git switch --create backport-65169-to-v9.4.x origin/v9.4.x
> # Cherry-pick the merged commit of this pull request and resolve the conflicts
> git cherry-pick -x d96b0a71d324abb637e8cd3559dc12f75d5d00e9
> # Push it to GitHub
> git push --set-upstream origin backport-65169-to-v9.4.x
> git switch main
> # Remove the local backport branch
> git branch -D backport-65169-to-v9.4.x
> ```
> 
> Then, create a pull request where the `base` branch is `v9.4.x` and the `compare`/`head` branch is `backport-65169-to-v9.4.x`.

Not going to backport to `v9.4.x` unless required.

**Pull request merged**